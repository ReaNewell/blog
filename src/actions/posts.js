import database from '../firebase/firebase';
import { firebase, storage } from '../firebase/firebase';

export const addPost = ({id, title, body, postDate, categories}) => ({
    type: "ADD_POST",
    post: {
        id,
        title,
        body,
        postDate,
        categories
    }
});
export const startAddPost = (postData = {}, image) => {
    return (dispatch) => {
        let currentPost;
        const {
            title = "",
            body = "",
            postDate = {
                day: new Date().getDate(),
                month: new Date().getMonth(),
                year: new Date().getFullYear()
            },
            categories = []
        } = postData;
        const post = { title, body, postDate, categories };

        return database.ref(`posts`).push(post).then((ref) => {
            dispatch(addPost({
                id: ref.key,
                ...post
            }));
            currentPost = ref.key;
        }).then(() => {
            if (image) {
                const fileName = image.name;
                const storageRef = storage.ref(`blogPictures/${fileName}`);
                const uploadTask = storageRef.put(image);
                let downloadURL;

                // Listen for state changes, errors, and completion of the upload.
                uploadTask.on(firebase.storage.TaskEvent.STATE_CHANGED, function(snapshot) {
                    // Get task progress, including the number of bytes uploaded and the total number of bytes to be uploaded
                    var progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
                    console.log('Upload is ' + progress + '% done');
                    switch (snapshot.state) {
                        case firebase.storage.TaskState.PAUSED: // or 'paused'
                        console.log('Upload is paused');
                        break;
                        case firebase.storage.TaskState.RUNNING: // or 'running'
                        console.log('Upload is running');
                        break;
                    }
                    }, function(error) {
    
                    }, function() {
                        // Upload completed successfully, now we can get the download URL
                        console.log('Upload complete.')
                        downloadURL = uploadTask.snapshot.downloadURL;
                        return database.ref(`posts/${currentPost}`).update({ 
                            postPicture: downloadURL,
                            postPictureName: fileName
                        })
                    });
            }
        })
    }
}

export const removePost = (id) => ({
    type: "REMOVE_POST",
    id
});
export const startRemovePost = (id, imageName) => {
    return (dispatch) => {
        return database.ref(`posts/${id}`).remove().then(() => {
            dispatch(removePost(id));
        }).then(() => {
            if (imageName) {
                const storageRef = storage.ref(`blogPictures/${imageName}`);
                storageRef.delete().then(() => {
                    console.log('Image deleted');
                }).catch((error) => {
                    console.log(error);
                });
            }
        });
    }
}

export const setPosts = (posts) => ({
    type: "SET_POSTS",
    posts
});
export const startSetPosts = () => {
    return (dispatch) => {
        return database.ref(`posts`).once('value').then((snapshot) => {
            let posts = [];

            snapshot.forEach((childSnapshot) => {
                posts.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setPosts(posts));
        });
    };
};

export const updatePost = (id, updates) => ({
    type: "UPDATE_POST",
    id,
    updates
});
export const startUpdatePost = (id, updates, currentPicture, picture) => {
    return (dispatch) => {
        return database.ref(`posts/${id}`).update({
            title: updates.title,
            body: updates.body
        }).then(() => {
            dispatch(updatePost(id, updates));
        }).then(() => {
            if (picture) {
                const fileName = picture.name;
                const storageRef = storage.ref(`/blogPictures/${fileName}`);
                const uploadTask = storageRef.put(picture);
                let downloadURL;

                // Listen for state changes, errors, and completion of the upload.
                uploadTask.on(firebase.storage.TaskEvent.STATE_CHANGED, function(snapshot) {
                // Get task progress, including the number of bytes uploaded and the total number of bytes to be uploaded
                var progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
                console.log('Upload is ' + progress + '% done');
                switch (snapshot.state) {
                    case firebase.storage.TaskState.PAUSED: // or 'paused'
                    console.log('Upload is paused');
                    break;
                    case firebase.storage.TaskState.RUNNING: // or 'running'
                    console.log('Upload is running');
                    break;
                }
                }, function(error) {

                }, function() {
                    // Upload completed successfully, now we can get the download URL
                    console.log('Upload complete.')
                    downloadURL = uploadTask.snapshot.downloadURL;
                    return database.ref(`posts/${id}`).update({ 
                        postPicture: downloadURL,
                        postPictureName: fileName
                    }).then(() => {
                        if (currentPicture && currentPicture !== fileName) {
                            return storage.ref(`/blogPictures/${currentPicture}`).delete();
                        }
                    });
                });
            }
        })
    }
}