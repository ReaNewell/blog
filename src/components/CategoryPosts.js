import React from 'react';
import SummarizedPost from './SummarizedPost';

class CategoryPosts extends React.Component {
    render() {
        return (
            <div className="recent-posts">
                <h2 className="recent-posts__title">{`Articles Related To ${this.props.category}`}</h2>
                {this.props.posts.map((post) => (
                    <SummarizedPost 
                        key = {post.id}
                        title = {post.title}
                        imageName = {post.postPictureName}
                        imagePath = {post.postPicture}
                        date = {post.postDate}
                    />
                ))}
            </div>
        )
    }
}

export default CategoryPosts;