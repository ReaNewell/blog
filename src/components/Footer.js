import React from 'react';

export const Footer = () => (
    <div className="footer">
        <img alt='site-logo' src='/images/rea-logo.png'/>
    </div>
);