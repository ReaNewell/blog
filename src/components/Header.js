import React from 'react';
import { Navigation } from './Navigation';

export const Header = () => (
    <div>
        <Navigation />
    </div>
);