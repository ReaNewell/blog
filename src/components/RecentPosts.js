import React from 'react';
import SummarizedPost from './SummarizedPost';

class RecentPosts extends React.Component {
    render() {
        return (
            <div>
                <SummarizedPost />
            </div>
        )
    }
}

export default RecentPosts;