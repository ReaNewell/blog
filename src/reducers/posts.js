const defaultPostsReducerState = [];

export default (state = defaultPostsReducerState, action) => {
    switch (action.type) {
        case "ADD_POST":
            return state.concat({ ...action.post })
        default:
            return state;
    }
}