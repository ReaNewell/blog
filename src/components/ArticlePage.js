import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import { Footer } from './Footer';
import { Header } from './Header';
import SummarizedPost from './SummarizedPost';
import SideBar from './SideBar';
import moreArticles from '../selectors/moreArticles';
import dateFormatter from '../tools/dateFormatter';

export class ArticlePage extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            html: this.props.post.body.replace(/^"(.*)"$/, '$1')
        }
    } 
    componentDidMount() {
        window.scroll(0, 0);

        // Selects this the body element of this component and assign the HTML value of the post to that element.
        const body = ReactDOM.findDOMNode(this).getElementsByClassName('article__body')[0];
        body.innerHTML = this.state.html;
    }
    componentDidUpdate() {
        window.scroll(0, 0);
    }
    render() {
        return (
            <div className="article-page">
                <Header />
                    <article className="article">
                        <h2 className="article__title">{this.props.post.title.toUpperCase()}</h2>
                        <p className="article__date">{dateFormatter(this.props.post.postDate.day, this.props.post.postDate.month, this.props.post.postDate.year)}</p>
                        {this.props.post.postPicture && <img className="article__image" src={this.props.post.postPicture} alt={this.props.post.postPictureName}/>}
                        <div className="article__body" />

                        <div className='share-buttons'>
                            <a className="share-buttons__share-button" href={`http://facebook.com/sharer.php?u=https%3A%2F%2Fblog.reanewell.com/posts/${this.props.post.title.replace(/ /g, "-")}`}>Facebook</a>
                            <a className="share-buttons__share-button" href={`https://twitter.com/intent/tweet?url=https%3A%2F%2Fblog.reanewell.com/${this.props.post.title.replace(/ /g, "-")}`}>Twitter</a>
                            <a className="share-buttons__share-button" href={`https://www.linkedin.com/shareArticle?mini=true&url=https%3A%2F%2Fblog.reanewell.com/${this.props.post.title.replace(/ /g, "-")}&title=${this.props.post.title}`}>LinkedIn</a>
                        </div>

                        <h2 className='article__header'>More Articles</h2>
                        {this.props.moreArticles.map((post) => (
                            <SummarizedPost 
                                key = {post.id}
                                title = {post.title}
                                imageName = {post.postPictureName}
                                imagePath = {post.postPicture}
                                date = {post.postDate}
                            />
                        ))}
                    </article>
                <SideBar />
                <Footer />
            </div>
        )
    }
}
const mapStateToProps = (state, props) => {
    const post = state.posts.find((post) => post.title === props.match.params.title.replace(/-/g, ' '));

    return ({
        post,
        moreArticles: moreArticles(state.posts, post.id, post.category)
    })
};

export default connect(mapStateToProps)(ArticlePage);