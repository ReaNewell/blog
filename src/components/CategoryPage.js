import React from 'react';
import { connect } from 'react-redux';
import categoryFilter from '../selectors/categoryFilter';
import { Footer } from './Footer';
import { Header } from './Header';

import SideBar from './SideBar';
import CategoryPosts from './CategoryPosts';

export const CategoryPage = (props) => {
    componentDidMount = () => {
        window.scroll(0, 0);
    }
    componentDidUpdate = () => {
        window.scroll(0, 0);
    }
    return(
        <div className="homepage">
            <Header />
            <CategoryPosts 
                category={props.category}
                posts={props.posts}
            />
            <SideBar />
            <Footer />
        </div>
    )
};
const mapStateToProps = (state, props) => ({
    category: props.match.params.category,
    posts: categoryFilter(state.posts, props.match.params.category)
});

export default connect(mapStateToProps)(CategoryPage);