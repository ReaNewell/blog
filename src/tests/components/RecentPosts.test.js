import React from 'react';
import { shallow } from 'enzyme';
import { RecentPosts } from '../../components/RecentPosts';
import posts from '../fixtures/posts';

test('should render RecentPosts component correctly', () => {
    const wrapper = shallow(<RecentPosts posts={posts} />);
    expect(wrapper).toMatchSnapshot();
});