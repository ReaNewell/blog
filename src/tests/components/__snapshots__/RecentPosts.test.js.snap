// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RecentPosts component correctly 1`] = `
<div
  className="recent-posts"
>
  <h2
    className="recent-posts__title"
  >
    Recent Articles
  </h2>
  <SummarizedPost
    key="221"
    post={
      Object {
        "body": "This is the first blog post.",
        "categories": Array [
          "javascript",
          "html",
        ],
        "id": "221",
        "postDate": Object {
          "day": 1,
          "month": 4,
          "year": 2018,
        },
        "postPicture": "postPictureNumberTwo",
        "postPictureName": "postPictureNameNumberTwo",
        "title": "Title One",
      }
    }
  />
  <SummarizedPost
    key="123"
    post={
      Object {
        "body": "This is the first blog post.",
        "categories": Array [
          "css",
          "html",
        ],
        "id": "123",
        "postDate": Object {
          "day": 1,
          "month": 2,
          "year": 2017,
        },
        "postPicture": "postPicture",
        "postPictureName": "postPictureName",
        "title": "Title One",
      }
    }
  />
</div>
`;
